# 3.4 Loss functions & evaluation metrics

Gradient Descent · Track 3 (Classical ML). Runnable companion — every output matches the lesson.

## Look at WHERE the classifier errs, not just how often

In [1]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, precision_score, recall_score

X, y = load_breast_cancer(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=0, stratify=y)
scaler = StandardScaler().fit(X_train)
clf = LogisticRegression(max_iter=5000).fit(scaler.transform(X_train), y_train)
X_test_s = scaler.transform(X_test)
y_pred = clf.predict(X_test_s)

print(confusion_matrix(y_test, y_pred))

[[40  2]
 [ 0 72]]


In [2]:
# malignant (label 0) is the class we care about catching
print("recall (malignant caught):", round(recall_score(y_test, y_pred, pos_label=0), 3))
print("precision (malignant):     ", round(precision_score(y_test, y_pred, pos_label=0), 3))

recall (malignant caught): 0.952
precision (malignant):      1.0


Accuracy of 0.982 hid a 95.2% recall on malignancy - the model misses ~1 in 20 cancers. To catch more, lower the threshold (precision will drop).